library/include/conversion/hipsparse_bsr2csr.h

```diff
@@ -0,0 +1,105 @@
+/*! \file */
+/* ************************************************************************
+ * Copyright (C) 2024 Advanced Micro Devices, Inc. All rights Reserved.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * ************************************************************************ */
+#pragma once
```

Comment:
I would recommend against using `#pragma once` in public headers, particularly if you have split your headers into a large number of files like this. GCC uses a linked-list to remember which files are guarded by `#pragma once`, which will make `#include <hipsparse/hipsparse.h>` an $O(n^2)$ operation, where $n$ is the number of hipsparse headers. See https://gcc.gnu.org/bugzilla/show_bug.cgi?id=58770.

A few years ago, I wrote a utility that can convert pragma once to include guards or vice versa. You can use it to switch:

```bash
python3 -m venv venv3
. venv3/bin/activate
pip install guardonce
once2guard -r library/include -p 'path -2 | upper | remove HIPSPARSE_ | prepend HIPSPARSE_' -l -s '#endif /* % */'
```

That will convert all the public headers. The change for this file would be:
```patch
diff --git a/library/include/conversion/hipsparse_bsr2csr.h b/library/include/conversion/hipsparse_bsr2csr.h
index 5b8422f..7a79e5d 100644
--- a/library/include/conversion/hipsparse_bsr2csr.h
+++ b/library/include/conversion/hipsparse_bsr2csr.h
@@ -21,7 +21,8 @@
  * THE SOFTWARE.
  *
  * ************************************************************************ */
-#pragma once
+#ifndef HIPSPARSE_CONVERSION_HIPSPARSE_BSR2CSR_H
+#define HIPSPARSE_CONVERSION_HIPSPARSE_BSR2CSR_H
 
 #ifdef __cplusplus
 extern "C" {
@@ -102,4 +103,6 @@ hipsparseStatus_t hipsparseZbsr2csr(hipsparseHandle_t         handle,
 
 #ifdef __cplusplus
 }
-#endif
\ No newline at end of file
+#endif
+
+#endif /* HIPSPARSE_CONVERSION_HIPSPARSE_BSR2CSR_H */
\ No newline at end of file
```

You can tweak [the pattern](https://github.com/cgmb/guardonce/blob/master/docs/PatternLanguage.md) as you see fit.

Reply:
Thanks. Fixed.